src/targets/gpu/device/topk.cpp

```diff
@@ -0,0 +1,84 @@
+#include "migraphx/gpu/device/shape.hpp"
+#include <migraphx/shape.hpp>
+#include <migraphx/argument.hpp>
+#include <migraphx/gpu/device/topk.hpp>
+#include <migraphx/gpu/device/tensor.hpp>
+#include <migraphx/gpu/device/launch.hpp>
+#include <migraphx/gpu/device/types.hpp>
+#include <migraphx/gpu/device/visit.hpp>
+#include <migraphx/gpu/device/heap.hpp>
+#include <migraphx/ranges.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+namespace device {
+
+argument topk(hipStream_t stream,
+              argument val_res,
+              argument ind_res,
+              argument arg,
+              int64_t k,
+              int64_t axis,
+              bool largest)
```

Comment:
This function should be templated with a compare function(only defined in the cpp file), so it doesnt check a boolean on every compare inside the gpu code:

```cpp
template<class Compare>
argument topk(hipStream_t stream,
              argument val_res,
              argument ind_res,
              argument arg,
              int64_t k,
              int64_t axis,
              Compare compare);
```

And then add two device functions that pass in the different compare:

```cpp
argument topk_largest(hipStream_t stream,
                      argument val_res,
                      argument ind_res,
                      argument arg,
                      int64_t k,
                      int64_t axis);
{
    return topk(stream, val_res, ind_res, arg, k, axis, std::less<>{});
}

argument topk_smalles(hipStream_t stream,
                      argument val_res,
                      argument ind_res,
                      argument arg,
                      int64_t k,
                      int64_t axis)
{
    return topk(stream, val_res, ind_res, arg, k, axis, std::greater<>{});
}
```

Then the compute method can check the `largest` flag and call the correct device function.

Reply:
fixed